crates/shared/src/http_client.rs

```diff
@@ -19,11 +23,141 @@ pub async fn response_body_with_size_limit(
     Ok(bytes)
 }
 
+#[derive(Debug, Clone)]
+pub struct RateLimitingStrategy {
+    drop_requests_until: Instant,
+    /// How many requests got rate limited in a row.
+    times_rate_limited: u64,
+    back_off_growth_factor: f64,
+    min_back_off: Duration,
+    max_back_off: Duration,
+}
+
+impl RateLimitingStrategy {
+    pub fn try_new(
+        back_off_growth_factor: f64,
+        min_back_off: Duration,
+        max_back_off: Duration,
+    ) -> Result<Self> {
+        ensure!(
+            back_off_growth_factor.is_normal(),
+            "back_off_growth_factor must be a normal f64"
+        );
+        ensure!(
+            back_off_growth_factor > 1.0,
+            "back_off_growth_factor needs to be greater than 1.0"
+        );
+        ensure!(
+            min_back_off <= max_back_off,
+            "min_back_off needs to be <= max_back_off"
+        );
+        Ok(Self {
+            drop_requests_until: Instant::now(),
+            times_rate_limited: 0,
+            back_off_growth_factor,
+            min_back_off,
+            max_back_off,
+        })
+    }
+}
+
+impl RateLimitingStrategy {
+    /// Resets back off and stops rate limiting requests.
+    pub fn response_ok(&mut self) {
+        self.times_rate_limited = 0;
+        self.drop_requests_until = Instant::now();
+    }
+
+    /// Calculates back off based on how often we got rate limited in a row.
+    fn get_current_back_off(&self) -> Duration {
+        let mut back_off = self.min_back_off;
+        for _ in 0..self.times_rate_limited {
+            // do this iteratively to reasonably prevent panics on overflow
+            back_off = back_off.mul_f64(self.back_off_growth_factor);
+            if back_off >= self.max_back_off {
+                return self.max_back_off;
+            }
+        }
+        back_off
+    }
+
+    /// Returns updated back off if no other thread increased it in the mean time.
+    pub fn response_rate_limited(&mut self, previous_rate_limits: u64) -> Option<Duration> {
+        if self.times_rate_limited != previous_rate_limits {
+            // Don't increase back off if somebody else already updated it in the meantime.
+            return None;
+        }
+
+        self.times_rate_limited += 1;
+        let new_back_off = self.get_current_back_off();
+        self.drop_requests_until = Instant::now() + new_back_off;
+        Some(new_back_off)
+    }
+
+    /// Returns number of times we got rate limited in a row if we are currently allowing requests.
+    pub fn times_rate_limited(&self, now: Instant) -> Option<u64> {
+        if self.drop_requests_until > now {
```

Comment:
Shouldn't this condition be:
```suggestion
        if now > self.drop_requests_until {
```

From what I understood, we should return `Some(_)` if we should rate limit, and `drop_requests_until` is the instant up until we should rate limit.

Reply:
It's actually correct this way albeit not so pretty. This is a protection against following scenario:
1. getting rate limited for some time
2. allowing requests after exceeding current back off time
3. send 2 requests in parallel
4. both return 429
5. increase back off time twice

To prevent increasing the back off twice we need to make sure no other request already increased the back off before us. This can be determined by comparing `self.times_rate_limited` from before and after the request.
If those values don't match, some other request already caused the back off to increase.

That means if we are currently being rate limited we don't even need the value and return `None` causing us to error out in the caller. That makes the API feel backwards and unintuitive but I found not better way to protect against that than simply locking the `Mutex` for the entire request which would prevent parallel requests entirely.

Reply:
This still does not solve the [ABA problem](https://en.wikipedia.org/wiki/ABA_problem) but I think this is something we can live with.

Reply:
No problem. I didn't refresh the page before posting my comment so I saw too late that you updated your comment already.